src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,205 @@
+                }
+                else
+                {
+                    bool unres = IsUnreservedCharacter(c);
+                    encodedLength += unres ? 1 : 3;
+                }
+            }
+
+            // Return the string if no encoding is required
+            if (value.Length == encodedLength)
+            {
+                return value;
+            }
+
+            // Encode
+            return string.Create(encodedLength, value, Encode);
+
+            static void Encode(Span<char> span, string s)
+            {
+                Span<byte> unicodeBytesBuffer = stackalloc byte[MaxUnicodeCharsReallocate * MaxUtf8BytesPerUnicodeChar];
+
+                var writePosition = 0;
+                for (var i = 0; i < s.Length; i++)
+                {
+                    var current = s[i];
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
+
+                        for (var count = 0; count < numberOfBytes; count++)
+                        {
+                            EscapeAsciiChar(ref span, ref writePosition, (char)unicodeBytesBuffer[count]);
+                        }
+                        i += unicodeCharCount - 1;
+                    }
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1);
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBytes[c / 8] >> (c % 8)) & 1) != 0;
+        }
+
+        // TODO(JamesNK): Remove WIP decode once we are certain Uri.UnescapeData works correctly
+#if Decode
+        public static string PercentDecode(string value)
```

Comment:
I did most of decode, then realized that I needed to worry about multi-byte characters, surrogate characters, UTF8 validation, etc. Ain't nobody got time for that.

Uri.UnescapeData should probably work fine for decoding. Leaving this here until we are sure it is not needed.

Reply:
Yes, I think using `Uri.UnescapeData` data should work as the only difference in the percent-encoding is the choice of characters that get percent-encoded.